datanode/bootstrap/bootstrap_server.go

```diff
@@ -0,0 +1,470 @@
+		if err != nil {
+			return nil, err
+		}
+		columns, err := p.diskStore.ListArchiveBatchVectorPartyFiles(req.Table, int(req.Shard), batchID, version, seq)
+		if err != nil {
+			return nil, err
+		}
+		vps := make([]*pb.VectorPartyMetaData, len(columns))
+		for j, colID := range columns {
+			vps[j] = &pb.VectorPartyMetaData{
+				ColumnID: uint32(colID),
+			}
+		}
+		batches[i] = &pb.BatchMetaData{
+			BatchID: int32(batchID),
+			Size:    uint32(size),
+			ArchiveVersion: &pb.ArchiveVersion{
+				ArchiveVersion: version,
+				BackfillSeq:    seq,
+			},
+			Vps: vps,
+		}
+	}
+
+	m.Batches = batches
+	m.Meta = &pb.TableShardMetaData_FactMeta{
+		FactMeta: &pb.FactTableShardMetaData{
+			HighWatermark: cutoff,
+			BackfillCheckpoint: &pb.BackfillCheckpoint{
+				RedoFileID:     redoFileID,
+				RedoFileOffset: redoFileOffset,
+			},
+		},
+	}
+
+	return m, nil
+}
+
+func (p *PeerDataNodeServerImpl) FetchVectorPartyRawData(req *pb.VectorPartyRawDataRequest, stream pb.PeerDataNode_FetchVectorPartyRawDataServer) error {
+	sessionInfo := &sessionInfo{
+		table:   req.Table,
+		shardID: req.Shard,
+		nodeID:  req.NodeID,
+	}
+	var err error
+
+	var timeElapsed int64
+	timeStart := utils.Now()
+
+	logInfoMsg(sessionInfo, "FetchVectorPartyRawData called", "batch", req.BatchID, "col", req.ColumnID)
+	defer func() {
+		if err == nil {
+			logInfoMsg(sessionInfo, "FetchVectorPartyRawData succeed", req.BatchID, "col", req.ColumnID, "timeused", timeElapsed)
+		} else {
+			logErrorMsg(sessionInfo, err, "FetchVectorPartyRawData failed", req.BatchID, "col", req.ColumnID)
+		}
+	}()
+
+	if err = p.validateRequest(req.SessionID, req.NodeID, req.Table, req.Shard); err != nil {
+		return err
+	}
+
+	t, err := p.metaStore.GetTable(req.Table)
+	if err != nil {
+		return err
+	}
+
+	var reader io.ReadCloser
+	if t.IsFactTable {
+		reader, err = p.diskStore.OpenSnapshotVectorPartyFileForRead(req.Table, int(req.Shard), int64(req.GetArchiveVersion().ArchiveVersion),
+			req.GetArchiveVersion().BackfillSeq, int(req.BatchID), int(req.ColumnID))
+	} else {
+		reader, err = p.diskStore.OpenSnapshotVectorPartyFileForRead(req.Table, int(req.Shard), int64(req.GetSnapshotVersion().RedoFileID),
+			req.GetSnapshotVersion().RedoFileOffset, int(req.BatchID), int(req.ColumnID))
+	}
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	vp := &pb.VectorPartyRawData{}
+	buf := make([]byte, chunkSize)
+	done := false
+	for !done {
+		n, err := reader.Read(vp.Chunk)
+		if err != nil {
+			if err == io.EOF {
+				done = true
+			}
+			return err
+		}
+		if n > 0 {
+			vp.Chunk = buf[:n]
+			if err = stream.Send(vp); err != nil {
+				return err
+			}
+		}
+	}
+	// in milsec
+	timeElapsed = utils.Now().Sub(timeStart).Nanoseconds() / 1000000
+
+	return nil
+}
+
+func (p *PeerDataNodeServerImpl) validateSessionSource(sessionID int64, nodeID string) error {
+	if sessionID == 0 {
```

Comment:
if it is sessionID is int63(),  it can be 0 right?

Reply:
yeah, will change